File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/symlink_safety.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""Forbid unsafe symlinks rule."""

from collections.abc import Iterable

import os
from pathlib import Path

from harness_check_rule import Finding, HarnessCheckRule

from .utils import is_safe_directory, relative, safe_walk, severity_for

class SymlinkSafetyRule(HarnessCheckRule):
    """Validate symlinkSafety check."""

    category = "symlinkSafety"

    def applies(self, manifest: dict) -> bool:
        """Check if this rule applies to the manifest."""
        section = manifest.get(self.category)
        if not isinstance(section, dict):
            return False
        return section.get("enabled", True) is not False

    def validate(self, root: Path, manifest: dict) -> Iterable[Finding]:
        """Validate symlinkSafety check."""
        section = manifest.get(self.category, {})
        if not isinstance(section, dict):
            return []
        params = section.get("parameters", {})
        if not isinstance(params, dict):
            return []
        allowed_pairs = params.get("allowedSymlinkPairs", [])
        if not isinstance(allowed_pairs, list):
            return []
        messages = section.get("messages", {})
        if not isinstance(messages, dict):
            return []
        allowed_set = frozenset(
            tuple(sorted((p[0], p[1]))) for p in allowed_pairs
            if isinstance(p, list) and len(p) == 2
        )
        scan_bases = (".claude", "docs", ".github", "AGENTS.md", "CLAUDE.md", "ARCHITECTURE.md")
        root_findings = [
            Finding(
                severity_for(manifest, self.category),
                self.category,
                messages.get("scanRootNotAllowed", "symlink scan root is not allowed: {path}").format(
                    path=relative(base_path)
                ),
            )
            for base in scan_bases
            if (base_path := root / base).is_symlink()
            and tuple(sorted((base_path.name, os.readlink(base_path).split("/")[-1]))) not in allowed_set
        ]
        file_findings = []
        for base in scan_bases:
            base_path = root / base
            if base_path.is_symlink() or not is_safe_directory(base_path):
                continue
            for path in safe_walk(base_path):
                if not path.is_symlink():
                    continue
                if path.parent == root and path.name in {"AGENTS.md", "CLAUDE.md"}:
                    target_name = os.readlink(path)
                    pair = tuple(sorted((path.name, target_name)))
                    if pair not in allowed_set:
                        file_findings.append(Finding(
                            severity_for(manifest, self.category),
                            self.category,
                            messages.get("fileNotAllowed", "symlink file is not allowed: {path}").format(
                                path=relative(path)
                            ),
                        ))
                else:
                    file_findings.append(Finding(
                        severity_for(manifest, self.category),
                        self.category,
                        messages.get("pathNotAllowed", "symlink path is not allowed: {path}").format(
                            path=relative(path)
                        ),
                    ))
        return root_findings + file_findings


RULE: HarnessCheckRule = SymlinkSafetyRule()

File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/fs/symlink-safety.ts
#!/usr/bin/env bun
// -*- coding: utf-8 -*-
import type {
  Finding,
  HarnessCheckRule,
  RuleContext,
} from "../harness-check-rule";

/**
 * Forbid unsafe symlinks. All scanned files and directories MUST NOT be symbolic
 * links, except symlinks within parameters.allowedSymlinkPairs (root contract symlinks).
 */
export const symlinkSafetyRule: HarnessCheckRule = {
  category: "symlinkSafety",
  applies(ctx: RuleContext): boolean {
    const section = ctx.manifest.raw.symlinkSafety;
    if (typeof section !== "object" || section === null) {
      return false;
    }
    const enabled = (section as { enabled?: unknown }).enabled;
    return enabled !== false;
  },
  validate(ctx: RuleContext): readonly Finding[] {
    const section = ctx.manifest.raw.symlinkSafety;
    if (typeof section !== "object" || section === null) {
      return [];
    }
    const params = ctx.readJsonObject(
      (section as Record<string, unknown>).parameters,
    );
    const allowedPairsList = params.allowedSymlinkPairs;
    const allowedSet = new Set<string>();
    if (Array.isArray(allowedPairsList)) {
      for (const pairItem of allowedPairsList) {
        if (Array.isArray(pairItem) && pairItem.length >= 2) {
          allowedSet.add(JSON.stringify([String(pairItem[0]), String(pairItem[1])].sort()));
        }
      }
    }
    const messages = ctx.readJsonObject(
      (section as Record<string, unknown>).messages,
    );
    const scanRootNotAllowedMsg =
      typeof messages.scanRootNotAllowed === "string"
        ? messages.scanRootNotAllowed
        : "symlink scan root is not allowed: {path}";
    const fileNotAllowedMsg =
      typeof messages.fileNotAllowed === "string"
        ? messages.fileNotAllowed
        : "symlink file is not allowed: {path}";
    const pathNotAllowedMsg =
      typeof messages.pathNotAllowed === "string"
        ? messages.pathNotAllowed
        : "symlink path is not allowed: {path}";
    const severity = ctx.severityOf("symlinkSafety");
    const findings: Finding[] = [];
    const scanBases = [
      ".claude",
      "docs",
      ".github",
      "AGENTS.md",
      "CLAUDE.md",
      "ARCHITECTURE.md",
    ];
    for (const base of scanBases) {
      if (ctx.isSymlink(base)) {
        const target = ctx.allowedRootContractTarget(base);
        if (target === null) {
          findings.push({
            severity,
            category: "symlinkSafety",
            message: scanRootNotAllowedMsg.replace("{path}", base),
            file: base,
            startLine: 1,
            startColumn: 1,
            endLine: 1,
            endColumn: 1,
          });
        }
        continue;
      }
      if (!ctx.isDirectory(base)) {
        continue;
      }
      const [files, walkerFindings] = ctx.walkDirectory(base);
      findings.push(...walkerFindings);
      for (const file of files) {
        if (ctx.isSymlink(file)) {
          const target = ctx.allowedRootContractTarget(file);
          if (target === null) {
            const isSpecialFile =
              file.split("/").length === 2 &&
              (file === ".claude/AGENTS.md" || file === ".claude/CLAUDE.md");
            findings.push({
              severity,
              category: "symlinkSafety",
              message: isSpecialFile
                ? fileNotAllowedMsg.replace("{path}", file)
                : pathNotAllowedMsg.replace("{path}", file),
              file,
              startLine: 1,
              startColumn: 1,
              endLine: 1,
              endColumn: 1,
            });
          }
        }
      }
    }
    return findings;
  },
};
